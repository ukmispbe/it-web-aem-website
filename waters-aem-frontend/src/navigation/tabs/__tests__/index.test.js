jest.mock("../../../scripts/fade-x.js");

import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import Fader from "../../../scripts/fade-x.js";
import Tabs from "../";

describe("Feature: Tabs Component", () => {
    const propsWithOutItems = {
        items: [],
        activeIndex: -1,
        onClick: jest.fn(),
        enableFading: false
    };

    const propsWithItems = {
        ...propsWithOutItems,
        activeIndex: 0,
        items: [
            {
                name: "Tab One"
            },
            {
                name: "Tab Two"
            }
        ]
    };
    
    beforeEach(() => {
        React.useEffect = jest.fn(f => f());
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    describe("Scenario: Rendering Tabs", () => {
        describe("When no tabs are provided", () => {
            it("Then it should match snapshot", () => {
                const json = renderer.create(<Tabs {...propsWithOutItems} />);
                
                expect(json).toMatchSnapshot();
            });
        });

        describe("When tabs are provided", () => {
            it("Then it should match snapshot", () => {
                const json = renderer.create(<Tabs {...propsWithItems} />);
                
                expect(json).toMatchSnapshot();
            });
        });
    });

    describe("Scenario: Fading Tabs", () => {
        const tabsRef = {
            current: {
                addEventListener: jest.fn(),
                removeEventListener: jest.fn()
            }
        }

        beforeAll(() => {
            React.useRef = jest.fn(() => tabsRef);
        });

        describe("When fading is disabled", () => {
            it("Then it should not initialize fading", () => {
                shallow(<Tabs {...propsWithOutItems} />);

                expect(Fader).not.toHaveBeenCalled();
                expect(tabsRef.current.addEventListener).not.toHaveBeenCalled();
            });
        });

        describe("When fading is enabled but there are no items", () => {
            it("Then it should not initialize fading", () => {
                shallow(<Tabs {...propsWithItems} />);

                expect(Fader).not.toHaveBeenCalled();
                expect(tabsRef.current.addEventListener).not.toHaveBeenCalled();
            });
        });

        describe("When fading is enabled and there are items", () => {
            it("Then it should initialize fading", () => {
                const mockProps = {
                    ...propsWithItems,
                    enableFading: true
                };

                shallow(<Tabs {...mockProps} />);

                expect(Fader).toHaveBeenCalled();
                expect(tabsRef.current.addEventListener).toHaveBeenCalled();
            });
        });
    });

    describe("Scenario: User Interaction", () => {
        describe("When tab is clicked", () => {
            it("Then it should call the click handler property", () => {
                const wrapper = shallow(<Tabs {...propsWithItems} />);

                wrapper.find("Tab").first().simulate("click");

                expect(propsWithItems.onClick).toHaveBeenCalled();
            });
        });
    });
});