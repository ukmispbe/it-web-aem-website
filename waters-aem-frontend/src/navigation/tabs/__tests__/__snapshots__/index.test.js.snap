// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature: Tabs Component Scenario: Rendering Tabs When no tabs are provided Then it should match snapshot 1`] = `
<div
  className="cmp-tabs-wrapper"
>
  <div
    className="cmp-tabs "
  />
</div>
`;

exports[`Feature: Tabs Component Scenario: Rendering Tabs When tabs are provided Then it should match snapshot 1`] = `
<div
  className="cmp-tabs-wrapper"
>
  <div
    className="cmp-tabs "
  >
    <div
      className="cmp-tabs__tab active"
      onClick={[Function]}
    >
      <span
        className="cmp-tabs__tab-label"
      >
        Tab One
      </span>
    </div>
    <div
      className="cmp-tabs__tab"
      onClick={[Function]}
    >
      <span
        className="cmp-tabs__tab-label"
      >
        Tab Two
      </span>
    </div>
  </div>
</div>
`;
